### PR TITLE
libs: remove mutex from lib

### PR DESCRIPTION
**Pull request by AyushKaul**
Removed mutex from libs/rand, for threads with independent randomness.

Fixes: #2956 

* [x] Updated all code comments where relevant
* [x] Wrote tests
* [x] Updated CHANGELOG_PENDING.md

### CODE REVIEW

File: libs/rand/random_test.go [hunk 1]
```diff
@@ -74,21 +71,6 @@ func testThemAll() string {
 	return out.String()
 }
 
-func TestRngConcurrencySafety(t *testing.T) {
```
Comment (melekes) [inline libs/rand/random_test.go]: why remove?

File: libs/rand/random.go [hunk 2]
```diff
@@ -17,9 +16,8 @@ const (
 // They all utilize math/rand's prng internally.
 //
 // All of the methods here are suitable for concurrent use.
-// This is achieved by using a mutex lock on all of the provided methods.
+// Different randomness each instance
```
Comment (melekes) [inline libs/rand/random.go]: ? not sure what are trying to say with this sentence

File: libs/rand/random.go [hunk 3]
```diff
@@ -17,9 +16,8 @@ const (
...
 type Rand struct {
-	sync.Mutex
 	rand *mrand.Rand
 }
```
Comment (melekes) [inline libs/rand/random.go]: what do you think about removing grand completely?

File: libs/rand/random.go [hunk 4]
```diff
@@ -141,9 +139,7 @@ func RandPerm(n int) []int {
 // Rand methods
 
 func (r *Rand) Seed(seed int64) {
```
Comment (melekes) [inline libs/rand/random.go]: we'll need to make sure we seed each instance of Rand throughout the code

File: libs/rand/random.go (hunk 3 above)
Comment (pushcodeveryday) [reply libs/rand/random.go]: rand is thread-safe by default. In the case of `NewSource()`, we can still use grand. But if we don't need global randomness then it's fine.

File: libs/rand/random.go (hunk 4 above)
Comment (pushcodeveryday) [reply libs/rand/random.go]: this is done for seeding:
```
for i := 0; i < 8; i++ {                                                     
         seed |= uint64(bz[i])                                                    
         seed <<= 8                                                               
}
```
you want to replace this?

Comment (melekes) [reply libs/rand/random.go]: we only need OS random in one / two places, in other we can just seed with the timestamp I believe

File: libs/rand/random_test.go (hunk 1 above)
Comment (pushcodeveryday) [reply libs/rand/random_test.go]: I have removed the mutex so this test did not make sense.

File: libs/rand/random.go (hunk 2 above)
Comment (pushcodeveryday) [reply libs/rand/random.go]: I'll have to re-phrase it.

File: libs/rand/random.go (hunk 3 above)
Comment (pushcodeveryday) [reply libs/rand/random.go]: I don't see how that can be of use. Do you have something on your mind?

File: libs/rand/random.go (hunk 4 above)
Comment (pushcodeveryday) [reply libs/rand/random.go]: okay. Let me see if I can help with that.

File: libs/rand/random.go (hunk 2 above)
Comment (pushcodeveryday) [reply libs/rand/random.go]: I have removed the comment.

**Comment by pushcodeveryday** [comment]
> what do you think about removing grand completely?

@melekes , I have tried to remove grand. Not sure if this is correct.

File: libs/rand/random.go [hunk 5]
```diff
@@ -54,96 +42,115 @@ func (r *Rand) reset(seed int64) {
 // Global functions
 
 func Seed(seed int64) {
```
Comment (melekes) [inline libs/rand/random.go]: I think the idea in the original issue was to get rid of these global functions completely. Instead, each "thread of execution" (i.e. caller) would have a `rand.New()` instance (probably as part of the main struct) and call `rand.XXX` (e.g. rand.Intn) each time it needs a random unit

Comment (pushcodeveryday) [reply libs/rand/random.go]: ok, got it.

**Comment by melekes** [comment]
Could you please rebase your PR against the latest master? Also, do you think you will have time to work on https://github.com/tendermint/tendermint/pull/4252#discussion_r362435151 ?

**Comment by pushcodeveryday** [comment]
I will be able to make the changes in a few days.

**Comment by melekes** [comment]
Closing due to inactivity. @AyushKaul please drop a comment here if you still want to finish this. Thank you!

**Pull request closed without merging**